### DISCUSSION TITLE
M1 MBP macOS Monterey (12.5.1 ), UR22 MK2, Blackhole - not as expected

### DISCUSSION THREAD

**Discussion by md1986** [jamulussoftware/jamulus#2795] [Technical Support & Help]
Hey guys,

I am working on a solution that could keep the band together even though our frontline singer has moved hundreds of kilometers away. The Jamulus server was easy enough to setup. The client somehow doesn't seem to be that easy though... I'm a bit desperate by now after trying a lot of things.

So what does work? I can connect to the server just fine and I also hear my sound input, both microphone and E-Bass. The caveat: this comes through only one channel into the server. The Yamaha driver for the UR22 MK2 in the MIDI settings also shows that it has 2 inputs. In GarageBand for example I can also select both inputs separately for a mic track and another E-Bass track. 

But when I select the audio interface in Jamulus I am not being presented with the audio mapping options as one would expect. Also using 2ch Blackhole as an input device does not change this behavior. Am I doing something wrong or is this a bug in the macOS version of the client?

What do I want to achieve? I want to have two channels running into the server, one for the mic and one for the E-Bass. I could accomplish that with starting up another instance of Jamulus but I suspect that this is not how it should be and it is also far from being user-friendly if not to say confusing - even with naming the clients.

Any help is much appreciated.

**Comment by henkdegroot** [comment]
Well desired or not desired, I believe the approach using two instances of Jamulus would work best. Also for the "other" users/clients. As this will allow them to also control the volume and pan of the voice separate from the E-Bass.

I don't have a macOS, so not able to provide much more help.

**Comment by pljones** [comment]
> I could accomplish that with starting up another instance of Jamulus but I suspect that this is not how it should be and it is also far from being user-friendly if not to say confusing - even with naming the clients.

As @henkdegroot says, this _is_ how it should be.  It's far easier for other users if each part of your mix is sent to the server separately.  They can then adjust it if needs be.  (Although, do make sure you're sending loud and clear on each instance.)  Each instance then also shows what your instrument is (mic/vocals or bass), so that also makes it easier to understand what's going on for other people.  You just need to run each instance with a separate `--inifile` parameter.

I don't know MacOS either but my suggested layout is to have the mono audio in from the mic and the MIDI in from the E-Bass both going to Garageband and then setting up two stereo out "send" tracks, one for the mic, one for the E-Bass audio, that you connect to Jamulus using virtual audio cable (Blackhole) and return the audio from each to "return" tracks.  Only _one_ of these should feed your master out, the other can be dropped (as they contain the same thing - the full server mix).

**Reply by md1986** [reply]
That's exactly the answer I was (not) hoping for. So there's nothing wrong with the setup. It turns out that Jamulus is simply not capable of opening two streams to the server at once and map the channels accordingly from one instance. There has been a pull request for a phantom client from one instance that was closed (I think I remember by yourself) without a comment some 5 months back. Would you mind to explain why? I think (and from what I've read so many times: I know for a fact) that having two channels run into the server from one client is a very basic feature request. Why make the user go through the hassle of having to deal with two instances, which is not straight-forward at all? I develop software for a living and I do not mind fiddling with some settings and reading some docs to get where I want. On the other end of the scale there's our lead singer that is a total plug-and-play guy who without help would never be able (and willing) to jump through these hoops. From my point of view it is just unnecessarily complicated at the moment - no offense.

**Reply by chrisrimple** [reply]
I'm unclear what you mean by "two streams into the server" when your ask earlier was for two channels to the server, one for mic and one for e-bass.  You can do that today, following the instructions I provided below.  In short, route mic to BlackHole channel 1 and e-bass to BlackHole channel 2, then set Jamulus to IN:BlackHole with inputs Channel 1 and Channel 2.

The recommendation to keep mic and e-bass in separate Jamulus instances is for the benefit of the other musicians.  If you run a single instance, any volume adjustments they make will affect mic AND e-bass; if you run two instances, you provide the other musicians with separate volume controls.

**Reply by md1986** [reply]
That's exactly the point. It should not be necessary to have two instances of Jamulus for that. It works with two instances but it feels clunky and unnatural. Especially on macOS where the default behavior is that there's only one instance of any program.
This is a very basic feature that probably 80+% of musicians need because they have a mic and an instrument. And of course you would want both channels to show up separately as two faders in the mixing console so the volume can be adjusted independently. I would like to understand why such a basic feature has not made it to Jamulus in years even though it has been requested over and over again and seemingly there was a (maybe imperfect; but at least it was a start and backwards-compatible) solution already. 

Aside from that I think I can get it to work with the tips you guys gave and I am very grateful for your help!

**Reply by pljones** [reply]
The point of Jamulus is to connect a single audio stream to a server and back (or to mix audio streams from a bunch of clients).  That's all.

If you want to send multiple audio streams -- simple: multiple Jamulus clients.  It's that easy.

**Reply by md1986** [reply]
Thanks for the explanation. I would not call it easy but agree to disagree. Once again thank you for your support. We will make it work with the provided knowledge.

**Reply by chrisrimple** [reply]
@md1986 I think you're extrapolating your needs as "the needs of most musicians", but are inaccurate.  Based on my experience over the last 2 years, including participation in plenty of open jams and the WorldJam via Jamulus, most musicians have a single instrument and do not sing.  If you were to look back at WorldJam performances, which have involved literally hundreds of musicians performing thousands of songs, perhaps 10% were both playing an instrument AND singing.  Yes, they may use a mic when talking with others between songs, but they don't need a mic while playing.  And for them, a single stream (with mic left and instrument right, or the reverse) is fine, because the other musicians are primarily concerned about mixing the instrument volume while playing together, and don't care that a single volume slider affects instrument + mic.

That said, even with a single stream you can still achieve relative volume control by using the Pan control, so that you're hearing more or less of instrument or mic, then use the volume slider to increase/decrease both.  For this to be available, the musician must be sending Stereo, not Mono (in Jamulus audio settings).

**Reply by md1986** [reply]
This is a very good comment! Thank you once again for your input. And I agree with you: I probably extrapolated and combined the need for talkback and actual vocals. I still think that another fader would be more straightforward than using a pan knob. But functionally it will definitely work.

**Comment by chrisrimple** [comment]
Spend some time here...

https://docs.google.com/document/d/1smcvsxdaaViPQvGMQHmah_6BQeqowhmGSFMHfnlY2FI/

GarageBand cannot be used with Jamulus, since Jamulus requires 48KHz and GarageBand only supports 44KHz.  If you need a DAW or other audio-generating software to be running for the e-bass, I recommend that you purchase MainStage ($30 USD).  Then you can route MIDI from the e-bass to MainStage for sound generation, also route your mic to MainStage, and then either (1) output together to BlackHole as Channel 1 and Channel 2, or output separately as Channels 1-4.  You'll need to run an instance of Jamulus for every 2 channels that you're outputting, and you need BlackHole as an intermediary because MainStage does not include its own virtual audio driver.

**Reply by md1986** [reply]
I found this document before but due to its sheer size I must've overlooked the remark about Garageband. Thanks for the heads up and the recommendation of MainStage. Looks like an awesome tool! I might even consider using it when performing live if it works well in this context.

**Reply by chrisrimple** [reply]
Many musicians use MainStage live, and it's hugely popular in the musical theater world.  If you decide to try it, there are some excellent learning resources on YouTube and in Facebook Groups.

**Reply by md1986** [reply]
I would like to use it in conjunction with my E-Bass. I’m not sure that the pedalboard will work with that. In GarageBand it mostly works for guitars only. So I suspect it will be similar in MainStage. Maybe I also need to invest in a plug-in called Bass Essentials 3 which adds a modeling amp with the most important built-in effects. Not 100% what I imagine though. But maybe I can find some good AU plug-ins for the E-Bass to add some effect pedals.

**Reply by chrisrimple** [reply]
MainStage is a live performance version of Apple's Logic Pro application, so has all the same instrument sounds available, and you can add VSTs like Bass Essentials.  If your e-bass is sending MIDI, MainStage can play any sound you want: harpsichord via bass?  Yes.  :)

**Comment by md1986** [comment]
I am currently trying to find my way around the .ini files but can't really seem to get it to work. I start the program with --inifile parameter from the terminal and get this output:
```
MBP-2021 ~ % /Applications/Jamulus.app/Contents/MacOS/Jamulus --inifile "/Users/$USER/Documents/Jamulus/mic.ini"
- initialization file name: /Users/$USER/Documents/Jamulus/mic.ini
- allocated port number: 22134
```

So it seems that it picks up the ini file but neither do I see preset values like the name, audio interface, etc., nor does it save anything to that .ini file when closing the app. Did I miss something?

**Reply by pljones** [reply]
A couple of diagnostic commands:
```
env | grep MBP
```
```
testvar=value; echo "testvar is $testvar"
```

The point here is that
```
- initialization file name: /Users/$USER/Documents/Jamulus/mic.ini
```
is wrong -- the `$USER` should have been replaced with `MBP-2021`.  Better might be to replace `/Users/$USER/` with `~/` and see if that works.  As the value `~` has no spaces in, you shouldn't need the quotes around it, either.  If that doesn't work...
```
cd
pwd
```
will tell you the value that `/Users/$USER` is meant to represent.  You should be able to use that instead.

This comes with the caveat that under very recent versions of MacOS, if I've understood what I've heard correctly, each app gets run in its own virtual session with limited access to the outside world.  It may turn out that's a problem, as well...

**Reply by md1986** [reply]
Sorry for the confusion. $USER is just a placeholder I replaced the original user name with. Should've made that more clear. 

I think your last paragraph might be the key. I did not see the app asking for permission to write to the Documents folder, which other apps usually do. It's also not possible to add that manually in the security section of macOS. The app has to request that on its own. Anything I can do from here to help out with troubleshooting or testing something?

**Reply by md1986** [reply]
Just for completeness sake:

I replaced the path to the user home with `~` as suggested. The result remains the same (as expected): no ini file is created.

```
myuser@my-MBP-2021 ~ % /Applications/Jamulus.app/Contents/MacOS/Jamulus --inifile "~/Documents/Jamulus/mic.ini"
- initialization file name: ~/Documents/Jamulus/mic.ini
- allocated port number: 22134
```

**Reply by pljones** [reply]
How about trying `/tmp` in place of `/Users/$USER/Documents/Jamulus/` and seeing if it works there?  Does `/Users/$USER/Documents/Jamulus/` exist?
